### PR TITLE
added a theme provider that watches a dir for new theme jars

### PR DESCRIPTION
**Pull request by xgp**

### CODE REVIEW

File: src/main/java/io/phasetwo/keycloak/themes/theme/JarFolderThemeProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  }
+
+  @Override
+  public ThemeProvider create(KeycloakSession session) {
+    return new JarFolderThemeProvider(session, globalThemes, realmThemes);
+  }
+
+  @Override
+  public void init(Config.Scope config) {
+    String d = config.get("dir");
```
[the remark this answers is not in the document]
Comment (xgp) [reply src/main/java/io/phasetwo/keycloak/themes/theme/JarFolderThemeProviderFactory.java:189]: Got it. I added an `EnvironmentDependentProviderFactory.isSupported` method to block the extension if the dir name isn't configured and actually a directory.

File: README.md [hunk 2]
```diff
@@ -94,6 +95,18 @@ The implementation of `EmailTemplateProvider` that allows the use of mustache.js
 - We get equivalent funcationlity to the methods like `linkExpirationFormatter(linkExpiration)` by using the library's lambda functionality, and using the mustache-y syntax `{{#linkExpirationFormatter}}{{linkExpiration}}{{/linkExpirationFormatter}}`, but there isn't complete coverage yet.
 - There is essentially no i18n at this point, so only the english templates work.
 
+### JAR Folder Theme Provider
+
+This includes an implementation of `ThemeProvider` that dynamically loads theme JARs from a specified directory at runtime. This is useful for deploying packaged themes without requiring a restart of Keycloak. The specified directory is scanned for JAR files both at the top level, and 1 directory deep. The JAR files placed in the top level will expose the enclosed themes to all realms. The subdirectories are meant to be named with realm names that should be allowed to access the JAR files contained in those subdirectories.
```
[the remark this answers is not in the document]
Comment (xgp) [reply README.md]: Thanks. I clarified this in the readme

File: README.md [hunk 3]
```diff
@@ -94,6 +95,18 @@ The implementation of `EmailTemplateProvider` that allows the use of mustache.js
...
+
+In order to workaround some issues with the `ThemeProvider` API, subdirectory themes are named as `<realmName>--<themeName>`. Also, subdirectory themes are **only** available in the Admin UI for selection when you are logged in to that realm's Admin UI, **not** the `master` Admin UI.
+
+The following variables can be set in order to configure this provider:
+|Variable|Required|Default|Description|
+|----|----|----|----|
+|`--spi-theme-cache-themes`|yes||Must be set to `true` in order to use this provider.|
```
[the remark this answers is not in the document]
Comment (xgp) [reply README.md]: Fixed.

File: src/main/java/io/phasetwo/keycloak/themes/theme/JarFolderThemeProviderFactory.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  protected static Map<Theme.Type, Map<String, JarFileSystemTheme>> globalThemes = new HashMap<>();
+  protected static Map<Path, FileSystem> jars = new HashMap<>();
+
+  private File rootDir;
+  private FileAlterationMonitor monitor;
+
+  @Override
+  public void onFileChange(File file) {
+    log.infof("onFileChange %s", file);
+    onFileModified(isSubDir(file), file.toPath());
```
[the remark this answers is not in the document]
Comment (xgp) [reply src/main/java/io/phasetwo/keycloak/themes/theme/JarFolderThemeProviderFactory.java]: Good catch. I think we have to "mock" the whole removal and addition process when a file changes. I added this.

**Comment by xgp** [comment]
@rtufisi I resolved the issues. Please take a look again.

**Review by rtufisi — APPROVED** [review]

**Pull request merged**